src/main/java/org/terasology/alterationEffects/ComponentBasedAlterationEffect.java

```diff
@@ -0,0 +1,195 @@
+     * alteration effect. If that is the case, the upsert operation only needs to handle component creation, and setting
+     * the correct state is handled by {@link #updateComponent(OnEffectModifyEvent, Component, EffectContext)}.
+     * <p>
+     * <strong>Example: </strong> A common implementation pattern is to first retrieve the component from the optional
+     * value or create a new instance, update any fields as necessary, and then return the component:
+     * <pre>
+     * {@code
+     * MyComponent upsertComponent(Optional<MyComponent> maybeComponent, final EffectContext context) {
+     *      MyComponent component = maybeComponent.orElse(new MyComponent());
+     *      // set fields of 'component'  as necessary
+     *      return component;
+     * }
+     * }
+     * </pre>
+     *
+     * @param maybeComponent the value of the component if it is already present, or empty otherwise
+     * @param context the contextual information this effect is applied in
+     * @return the initialized or updated component
+     */
+    protected abstract C upsertComponent(Optional<C> maybeComponent, EffectContext context);
+
+    /**
+     * Hook method called if the backing component should be updated based on the initial {@link OnEffectModifyEvent}.
+     * <p>
+     * This hook is only called when the initial {@link OnEffectModifyEvent} is not consumed and both duration and
+     * magnitude modifiers are set. The event is sent out with a base magnitude of 0 and a base duration of 0. Any
+     * system can modify the event to change these values.
+     * <p>
+     * <strong>Note:</strong> Ideally, the corresponding system should act upon the {@link OnEffectModifyEvent} if the
+     * backing component is present on an entity. This should be the main method to update the entities state. If the
+     * corresponding system does react on the event, the initial {@link #upsertComponent(Optional, EffectContext)} can
+     * be limited to creating the component.
+     *
+     * @param event event holding modifications to duration or magnitude by other systems
+     * @param component the component to update
+     * @param context the contextual information this effect is applied in
+     * @return the updated component
+     */
+    protected abstract C updateComponent(OnEffectModifyEvent event, C component, EffectContext context);
+
+    /**
+     * Hook method to be called if the backing component should be removed without applying the effect.
+     * <p>
+     * Called for
+     * <ul>
+     *     <li>indefinite effects without modifiers</li>
+     * </ul>
+     * and
+     * <ul>
+     *     <li>definite effects with
+     *         <ul>
+     *             <li>invalid modifiers</li>
+     *             <li>no modifiers and zero duration</li>
+     *             <li>no modifiers and invalid duration</li>
+     *         </ul>
+     *     </li>
+     * </ul>
```

Comment:
This part is contributed by @jdrueckert - I take no credits for this (that's what goes beyon my knowledge, I cannot wrap around my head around this conditional construct 🙈 )